atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientImpl.java

```diff
@@ -0,0 +1,336 @@
+        this.clientName = clientName;
+        this.cluster = cluster;
+        this.session = session;
+        this.executor = executor;
+        this.healthCheckStatement = healthCheckStatement;
+    }
+
+
+    @Override
+    public String sessionName() {
+        return clientName;
+    }
+
+    @Override
+    public void close() {
+        healthCheckExecutor.shutdownNow();
+        log.info("Shutting down health checker for CQL client {}", SafeArg.of("cqlClient", clientName));
+        boolean shutdown = false;
+        try {
+            shutdown = healthCheckExecutor.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            log.error("Interrupted while shutting down health checker for CQL client {}. Should not happen",
+                    SafeArg.of("cqlClient", clientName));
+            Thread.currentThread().interrupt();
+        }
+
+        if (!shutdown) {
+            log.error("Failed to shutdown health checker in a timely manner for {}",
+                    SafeArg.of("clusterSession", clientName));
+        } else {
+            log.info("Shut down health checker for {}", SafeArg.of("cqlClient", clientName));
+        }
+    }
+
+    private CqlClientImpl start() {
+        healthCheckExecutor.scheduleAtFixedRate(() -> {
+            ListenableFuture<String> time = getTimeAsync();
+            try {
+                log.info("Current cluster time is: {}, for client {}",
+                        SafeArg.of("clusterTime", time.get()),
+                        SafeArg.of("cqlClient", clientName));
+            } catch (Exception e) {
+                log.info("CQL client health check failed for: {}", SafeArg.of("cqlClient", clientName));
+            }
+        }, 0, 1, TimeUnit.MINUTES);
+        return this;
+    }
+
+    // TODO (OStevan): make a prepared statement cache as per
+    //  https://docs.datastax.com/en/developer/java-driver/3.7/manual/statements/prepared/
+    @Override
+    public PreparedStatement prepareStatement(String queryString) {
+        return session.prepare(queryString);
+    }
+
+    @Override
+    public <V, R> ListenableFuture<R> executeQuery(CqlQuery<V, R> cqlQuery) {
+        Visitor<V, R> visitor = cqlQuery.createVisitor();
+        return transform(Futures.transformAsync(session.executeAsync(cqlQuery.boundStatement()), iterate(visitor),
+                executor), Visitor::result);
+    }
+
+    @Override
+    public <V, P, R> ListenableFuture<R> executeQueries(Stream<CqlQuery<V, P>> inputStatementPairStream,
+            Function<List<P>, R> transformer) {
+        List<ListenableFuture<P>> allResults = inputStatementPairStream
+                .map(this::executeQuery)
+                .collect(Collectors.toList());
+
+        return transform(Futures.allAsList(allResults), transformer);
+    }
+
+    private ListenableFuture<String> getTimeAsync() {
+        return Futures.transform(session.executeAsync(healthCheckStatement.bind()),
```

Comment:
your health check statement does not use your api, but uses the driver api directly, this makes it difficult for me to see how to actually use your api, and how to create a `CqlQuery` etc.